File: src/main/agent.ts
/**
 * 로그인 시 자동 실행 (launchd).
 *
 * 앱이 데몬을 품고 있으므로 앱 자신을 등록한다. 개발 중에는 electron 이
 * 프로젝트를 여는 형태라 등록해도 의미가 없으니 그때는 막는다.
 */

import { execFile } from 'node:child_process'
import { existsSync, mkdirSync, rmSync, writeFileSync } from 'node:fs'
import { homedir, userInfo } from 'node:os'
import { dirname, join } from 'node:path'
import { promisify } from 'node:util'
import { app } from 'electron'

const run = promisify(execFile)

export const LABEL = 'com.jkang.mirabox'
export const PLIST_PATH = join(homedir(), 'Library', 'LaunchAgents', `${LABEL}.plist`)
export const LOG_DIR = join(homedir(), 'Library', 'Logs', 'mirabox')

const domain = () => `gui/${userInfo().uid}`

async function launchctl(...args: string[]): Promise<boolean> {
  try {
    await run('launchctl', args)
    return true
  } catch {
    return false
  }
}

export const isPackaged = () => app.isPackaged
export const isInstalled = () => existsSync(PLIST_PATH)
export const isRunning = () => launchctl('print', `${domain()}/${LABEL}`)

function plist(): string {
  const escape = (text: string) =>
    text.replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;')
  const args = [app.getPath('exe')].map((a) => `      <string>${escape(a)}</string>`).join('\n')
  return `<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
  <dict>
    <key>Label</key><string>${LABEL}</string>
    <key>ProgramArguments</key>
    <array>
${args}
    </array>
    <key>RunAtLoad</key><true/>
    <key>ProcessType</key><string>Interactive</string>
    <key>StandardOutPath</key><string>${escape(join(LOG_DIR, 'agent.log'))}</string>
    <key>StandardErrorPath</key><string>${escape(join(LOG_DIR, 'agent.err'))}</string>
  </dict>
</plist>
`
}

export async function install(): Promise<void> {
  mkdirSync(LOG_DIR, { recursive: true })
  mkdirSync(dirname(PLIST_PATH), { recursive: true })
  writeFileSync(PLIST_PATH, plist(), 'utf8')
  await launchctl('bootout', domain(), PLIST_PATH)
  await launchctl('bootstrap', domain(), PLIST_PATH)
}

export async function uninstall(): Promise<void> {
  await launchctl('bootout', domain(), PLIST_PATH)
  rmSync(PLIST_PATH, { force: true })
}

File: src/mirabox/core/agent.py
"""로그인 시 자동 실행 (launchd).

에이전트는 화면 없는 데몬을 돌린다. 조작 화면을 열면 같은 기기를 두고
다투게 되므로, 화면이 뜨는 동안에는 에이전트를 잠시 내렸다가 닫을 때
되살린다. 기기는 한 프로세스만 점유할 수 있다.
"""

from __future__ import annotations

import os
import plistlib
import subprocess
import sys
from pathlib import Path

LABEL = "com.jkang.mirabox"
PLIST_PATH = Path.home() / "Library" / "LaunchAgents" / f"{LABEL}.plist"
LOG_DIR = Path.home() / "Library" / "Logs" / "mirabox"


def _domain() -> str:
    return f"gui/{os.getuid()}"


def _daemon_command() -> list[str]:
    """설치된 콘솔 스크립트를 먼저 쓰고, 없으면 현재 인터프리터로 모듈을 부른다."""
    script = Path(sys.executable).with_name("mirabox")
    if script.exists():
        return [str(script)]
    return [sys.executable, "-m", "mirabox.core.daemon"]


def _launchctl(*args: str) -> subprocess.CompletedProcess:
    return subprocess.run(["launchctl", *args], capture_output=True, text=True)


def is_installed() -> bool:
    return PLIST_PATH.exists()


def is_running() -> bool:
    return _launchctl("print", f"{_domain()}/{LABEL}").returncode == 0


def install() -> None:
    LOG_DIR.mkdir(parents=True, exist_ok=True)
    PLIST_PATH.parent.mkdir(parents=True, exist_ok=True)
    plist = {
        "Label": LABEL,
        "ProgramArguments": _daemon_command(),
        "RunAtLoad": True,
        "KeepAlive": {"SuccessfulExit": False},
        "StandardOutPath": str(LOG_DIR / "daemon.log"),
        "StandardErrorPath": str(LOG_DIR / "daemon.err"),
        "ProcessType": "Background",
    }
    PLIST_PATH.write_bytes(plistlib.dumps(plist))
    start()


def uninstall() -> None:
    stop()
    PLIST_PATH.unlink(missing_ok=True)


def start() -> bool:
    if not PLIST_PATH.exists():
        return False
    _launchctl("bootout", _domain(), str(PLIST_PATH))     # 이미 떠 있으면 내린다
    return _launchctl("bootstrap", _domain(), str(PLIST_PATH)).returncode == 0


def stop() -> bool:
    return _launchctl("bootout", _domain(), str(PLIST_PATH)).returncode == 0


class Suspension:
    """조작 화면이 떠 있는 동안 에이전트를 내려 둔다."""

    def __init__(self) -> None:
        self._was_running = False

    def __enter__(self) -> "Suspension":
        self._was_running = is_installed() and is_running()
        if self._was_running:
            stop()
        return self

    def __exit__(self, *_exc) -> None:
        if self._was_running:
            start()


def main() -> None:
    """uv run mirabox-agent [install|uninstall|status]"""
    command = sys.argv[1] if len(sys.argv) > 1 else "status"
    if command == "install":
        install()
        print(f"등록 완료: {PLIST_PATH}")
        print(f"실행 명령: {' '.join(_daemon_command())}")
        print(f"로그: {LOG_DIR}")
    elif command == "uninstall":
        uninstall()
        print("해제 완료")
    else:
        print(f"등록됨: {'예' if is_installed() else '아니오'}")
        print(f"실행중: {'예' if is_running() else '아니오'}")
        if is_installed():
            print(f"plist: {PLIST_PATH}")


if __name__ == "__main__":
    main()
